src/components/js-tracer/Canvas.js

```diff
@@ -126,12 +128,25 @@ class JsTracerCanvasImpl extends React.PureComponent<Props, State> {
       );
     }
 
-    // Set the font size before creating a text measurer.
+    // Set the font before creating the text renderer. The font property resets
+    // automatically whenever the canvas size is changed, so we set it on every
+    // call.
     ctx.font = `${FONT_SIZE * cssToDeviceScale}px sans-serif`;
 
+    // Ensure the text measurement tool is created, since this is the first time
+    // this class has access to a ctx. We also need to recreate it when the scale
+    // changes because we are working with device coordinates.
+    if (
+      !this._textMeasurement ||
+      this._textMeasurementCssToDeviceScale !== cssToDeviceScale
+    ) {
+      this._textMeasurement = new TextMeasurement(ctx);
+      this._textMeasurementCssToDeviceScale = cssToDeviceScale;
+    }
```

Comment:
optional suggestion: use memoize-one with a function taking `cssToDeviceScale` as parameter. This would move the logic to invalidate the text measurement tool out of this function.

Reply:
Also this could possibly be handled directly by the shared Canvas code, passed to `drawCanvas` with the other properties. Because it looks like that the same code is repeated in all these charts.

Reply:
I agree, these suggestions would improve things! I'll leave them for follow-ups though.